// 0g/dist-esm/EntityImpostor.js
var Entity = /** @class */ (function () {
    function Entity() {
        var _this = this;
        this._id = 0;
        // TODO: make array
        this.components = new Map();
        this._persisted = false;
        // TODO: hide these behind Symbols?
        this.__set = function (entityId, components) {
            _this._id = entityId;
            _this.components.clear();
            components.forEach(function (comp) {
                _this.components.set(comp.type, comp);
            });
        };
        this.__addComponent = function (instance) {
            _this.components.set(instance.type, instance);
        };
        this.__removeComponent = function (typeId) {
            var instance = _this.components.get(typeId);
            _this.components.delete(typeId);
            return instance;
        };
        this.get = function (Type) {
            return _this.components.get(Type.id);
        };
        this.maybeGet = function (Type) {
            var _a;
            return ((_a = _this.components.get(Type.id)) !== null && _a !== void 0 ? _a : null);
        };
    }
    Object.defineProperty(Entity.prototype, "id", {
        get: function () {
            return this._id;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Entity.prototype, "persisted", {
        get: function () {
            return this._persisted;
        },
        enumerable: false,
        configurable: true
    });
    Entity.prototype.reset = function () {
        this.components.clear();
        this._persisted = false;
        this._id = 0;
    };
    Entity.prototype.clone = function (other) {
        var _this = this;
        other.components.forEach(function (value, key) {
            _this.components.set(key, value);
        });
        this._id = other.id;
    };
    Entity.prototype.persist = function () {
        this._persisted = true;
    };
    return Entity;
}());
export { Entity };
