// 0g/dist/__tests__/integration.test.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Component_1 = require("../Component");
var Effect_1 = require("../Effect");
var filters_1 = require("../filters");
var Game_1 = require("../Game");
var logger_1 = require("../logger");
var System_1 = require("../System");
var delta = 16 + 2 / 3;
describe('integration tests', function () {
    var OutputComponent = /** @class */ (function (_super) {
        __extends(OutputComponent, _super);
        function OutputComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return OutputComponent;
    }(Component_1.Component(function () { return ({
        removablePresent: false,
    }); })));
    var RemovableComponent = /** @class */ (function (_super) {
        __extends(RemovableComponent, _super);
        function RemovableComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return RemovableComponent;
    }(Component_1.Component(function () { return ({
        stepsSinceAdded: 0,
    }); })));
    var stepsTillToggle = 3;
    var SetFlagEffect = Effect_1.makeEffect([RemovableComponent, OutputComponent], function (ent) {
        logger_1.logger.debug('Setting removablePresent: true');
        ent.get(OutputComponent).update(function (output) {
            output.removablePresent = true;
        });
        return function () {
            logger_1.logger.debug('Setting removablePresent: false');
            ent.get(OutputComponent).update(function (output) {
                output.removablePresent = false;
            });
        };
    });
    var ReAddRemovableEffect = Effect_1.makeEffect([filters_1.not(RemovableComponent)], function (ent, game) {
        logger_1.logger.debug('Adding RemovableComponent');
        game.add(ent.id, RemovableComponent);
    });
    var IncrementRemoveTimerSystem = System_1.makeSystem([RemovableComponent], function (ent) {
        logger_1.logger.debug('Incrementing stepsSinceAdded');
        ent.get(RemovableComponent).update(function (comp) {
            comp.stepsSinceAdded++;
            logger_1.logger.debug("stepsSinceAdded: " + comp.stepsSinceAdded);
        });
    });
    var RemoveSystem = System_1.makeSystem([filters_1.changed(RemovableComponent)], function (ent, game) {
        if (ent.get(RemovableComponent).stepsSinceAdded >= stepsTillToggle) {
            logger_1.logger.debug('Removing RemovableComponent');
            game.remove(ent.id, RemovableComponent);
        }
    });
    it('adds and removes components, and queries for those operations', function () {
        var game = new Game_1.Game({
            components: [OutputComponent, RemovableComponent],
            systems: [
                SetFlagEffect,
                ReAddRemovableEffect,
                IncrementRemoveTimerSystem,
                RemoveSystem,
            ],
        });
        var a = game.create();
        game.add(a, OutputComponent);
        logger_1.logger.debug('Step 1');
        game.step(delta);
        var entity = game.get(a);
        expect(entity.maybeGet(OutputComponent)).not.toBe(null);
        expect(entity.get(OutputComponent).removablePresent).toBe(false);
        logger_1.logger.debug('Step 2');
        game.step(delta);
        entity = game.get(a);
        expect(entity.get(OutputComponent).removablePresent).toBe(true);
        expect(entity.maybeGet(RemovableComponent)).not.toBe(null);
        expect(entity.maybeGet(RemovableComponent).stepsSinceAdded).toBe(0);
        logger_1.logger.debug('Step 3');
        game.step(delta);
        entity = game.get(a);
        expect(entity.get(OutputComponent).removablePresent).toBe(true);
        expect(entity.maybeGet(RemovableComponent)).not.toBe(null);
        expect(entity.maybeGet(RemovableComponent).stepsSinceAdded).toBe(1);
        logger_1.logger.debug('Step 4');
        game.step(delta);
        entity = game.get(a);
        expect(entity.get(OutputComponent).removablePresent).toBe(true);
        expect(entity.maybeGet(RemovableComponent)).not.toBe(null);
        expect(entity.maybeGet(RemovableComponent).stepsSinceAdded).toBe(2);
        logger_1.logger.debug('Step 5');
        game.step(delta);
        entity = game.get(a);
        expect(entity.get(OutputComponent).removablePresent).toBe(false);
        expect(entity.maybeGet(RemovableComponent)).toBe(null);
        logger_1.logger.debug('Step 6');
        game.step(delta);
        entity = game.get(a);
        expect(entity.get(OutputComponent).removablePresent).toBe(true);
        expect(entity.maybeGet(RemovableComponent)).not.toBe(null);
        expect(entity.maybeGet(RemovableComponent).stepsSinceAdded).toBe(0);
    });
});
