// 0g/dist/filters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isChangedFilter = exports.isHasFilter = exports.isNotFilter = exports.isFilter = exports.changed = exports.not = exports.has = void 0;
var has = function (Component) { return ({
    Component: Component,
    kind: 'has',
    __isFilter: true,
    toString: function () {
        return "has(" + Component.name + ")";
    },
}); };
exports.has = has;
var not = function (Component) { return ({
    Component: Component,
    kind: 'not',
    __isFilter: true,
    toString: function () {
        return "not(" + Component.name + ")";
    },
}); };
exports.not = not;
var changed = function (Component) { return ({
    Component: Component,
    kind: 'changed',
    __isFilter: true,
    toString: function () {
        return "changed(" + Component.name + ")";
    },
}); };
exports.changed = changed;
var isFilter = function (thing) {
    return thing.__isFilter === true;
};
exports.isFilter = isFilter;
var isNotFilter = function (fil) {
    return fil.kind === 'not';
};
exports.isNotFilter = isNotFilter;
var isHasFilter = function (fil) {
    return fil.kind === 'has';
};
exports.isHasFilter = isHasFilter;
var isChangedFilter = function (fil) { return fil.kind === 'changed'; };
exports.isChangedFilter = isChangedFilter;
